pkg/handler/handler.go

```diff
@@ -140,17 +144,36 @@ func RegisterOpenAPIVersionedService(spec *spec.Swagger, servePath string, handl
 	return o, o.RegisterOpenAPIVersionedService(servePath, handler)
 }
 
+type ETagCache struct {
+	mutex        sync.Mutex
+	etag         string
+	lastModified time.Time
+}
+
+func (c ETagCache) LastModified(etag string) time.Time {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if c.etag == etag {
+		return c.lastModified
+	}
+	c.etag = etag
+	c.lastModified = time.Now()
+	return c.lastModified
+}
+
 // RegisterOpenAPIVersionedService registers a handler to provide access to provided swagger spec.
 func (o *OpenAPIService) RegisterOpenAPIVersionedService(servePath string, handler common.PathHandler) error {
 	accepted := []struct {
 		Type           string
 		SubType        string
-		GetDataAndETag func() ([]byte, string, time.Time, error)
+		GetDataAndETag func() ([]byte, string, error)
 	}{
 		{"application", "json", o.getSwaggerBytes},
 		{"application", "com.github.proto-openapi.spec.v2@v1.0+protobuf", o.getSwaggerPbBytes},
```

Comment:
Shouldn't these each have a separate etag cache?

Reply:
Because of the Vary header, I don't think that matters, clients can't re-use etags across content-type.

Reply:
I asked a more specific question below.